src/data/nat/strong_recursion.lean

```diff
@@ -0,0 +1,99 @@
+/-
+Copyright (c) 2020 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+
+import data.nat.basic
+import tactic
+
+/-!
+# Strong recursion
+
+A strong recursion principle based on `fin`.
+The benefit of `(Π (m:fin n), X m)` over `Π (m:ℕ) (h:m < n), X m`
+is that with `fin` the bound on `m` is bundled,
+and this can be used in later proofs and constructions.
+
+## Example
+
+For example, one can use this to give a definition of the Bernoulli numbers
+that closely follows the recursive definition found at
+https://en.wikipedia.org/wiki/Bernoulli_number#Recursive_definition
+It is:
+$$ B_n = \sum_{k < n} \binom{n}{k} \frac{B_k}{n - k + 1} $$
+
+```lean
+example : ℕ → ℚ :=
+strong_recursion $ λ n bernoulli,
+1 - finset.univ.sum (λ k : fin n, (n.choose k) * bernoulli k / (n - k + 1))
+```
+
+This example shows the benefit of using `fin n` in the implementation,
+because it allows us to use `finset.univ.sum` which enables the recursive call `bernoulli k`.
+If, on the other hand, we had used `Π (m:ℕ) (h:m < n), X m` and `(finset.range n).sum`,
+then the recursive call to `bernoulli k` would get stuck,
+because it would come with a proof obligation `k < n`
+which isn't provided by `(finset.range n).sum`.
+
+-/
+
+namespace nat
+universe variable u
+variables {X : ℕ → Sort u} (f : Π n, (Π (m:fin n), X m) → X n)
+
+/-- An auxilliary definition for `nat.strong_recursion`. -/
```

Comment:
```suggestion
/-- An auxiliary definition for `nat.strong_recursion`. -/
```
Also, saying very briefly what the definition does would be helpful, because it is a little bit hard to parse :)

Reply:
Good point. Done.